### DISCUSSION TITLE
How can I get the segmentation results as a separate image

### DISCUSSION THREAD

**Discussion by abdelmageed95** [open-edge-platform/anomalib#789] [Q&A] [answered]

**Comment by djdameln** [comment, marked as answer]
We currently do not have a mechanism for saving the segmentation results as separate images, so you will have to do some manual implementation for this. The easiest way to achieve this would be to retrieve the segmentation masks from the predictions of the PyTorch Lightning trainer instance. You could use to our [inference entrypoint script](https://github.com/openvinotoolkit/anomalib/blob/main/tools/inference/lightning_inference.py) as an example of how to make predictions with the PL trainer.

Just run the inference on your dataset and assign the output of `trainer.predict(...)` to a new variable. This variable will contain all prediction results of your input data, including the predicted segmentation masks. More specifically, replace [this line](https://github.com/openvinotoolkit/anomalib/blob/main/tools/inference/lightning_inference.py#L71) of the inference script with the following content, and add your own code to save the images to the file system:

```python
predictions = trainer.predict(model=model, dataloaders=[dataloader])
for pred in predictions:
    pred_mask = pred["pred_masks"]
    # save contents of pred_mask to an image file
    ...
```

**Reply by NoamKaufmanLin** [reply]
Can I get the heatmap via the same way? How so?

**Reply by djdameln** [reply]
The anomaly maps can be obtained in the same way by using the `anomaly_maps` key. The returned tensor will contain the raw values of the pixel-level anomaly score predictions, so you will need to perform some additional post-processing (e.g. `cv2.applyColorMap`) to convert these into RGB heatmaps.
